share/wake/lib/core/option.wake

```diff
@@ -17,100 +17,200 @@ package wake
-#   returning the first Some or a None
-# (fn: a => Option b) => List a => Option b
-export def findSomeFn fn =
-  def helper = match _
-    Nil = None
-    h, t = match (fn h)
-      Some x = Some x
-      None = helper t
-  helper
-
-# If a List of Options is all Some, returns Some of the elements, otherwise None
-# List (Option a) => Option (List a)
-export def findNone = findNoneFn (_)
-
-# Applies a function that returns Option to each element of a List
-#   returning None if any result is None, otherwise aggregates the Somes into a List
-# (fn: a => Option b) => List a => Option (List b)
-export def findNoneFn fn =
-  def helper = match _
-    Nil = Some Nil
-    h, t = match (fn h)
-      None = None
-      Some x = match (helper t)
-        Some t = Some (x, t)
-        None   = None
-  helper
+    Some a
+    None
+
+# isSome: Report if an Option has a value.
+# If you find yourself using the function, consider using a match instead.
+#
+#   isSome (Some 451) = True
+#   isSome (Some "x") = True
+#   isSome None       = False
+export def isSome: Option a => Boolean = match _
+    Some _ = True
+    None   = False
+
+# isSome: Report if an Option has no value.
+# If you find yourself using the function, consider using a match instead.
+#
+#   isSome (Some 451) = False
+#   isSome (Some "x") = False
+#   isSome None       = True
+export def isNone: Option a => Boolean = match _
+    Some _ = True
+    None   = False
+
+# getOrElse: extract the value from an Option, with a supplied default if None.
+# The default value expression is evaluated whether or not the Option is None.
+#
+#   int "not-an-integer"     # Option Integer
+#   | getOrElse 5            # Integer
+#   ---
+#   5
+export def getOrElse (default: a): Option a => a = match _
+    Some x = x
+    None   = default
+
+# getOrElse: extract the value from an Option, with a supplied default function if None.
+# The default value function is evaluated only when the Option is None.
+#
+#   int "567"
+#   | getOrElseFn (\Unit firstFactorOf 8947289472892423423423423235325)
+#   ---
+#   567
+export def getOrElseFn (fn: Unit => a): Option a => a = match _
+    Some a = a
+    None   = fn Unit
+
+# orElse: combine two Options, using the first value found, if any.
+#
+#   int "not-an-integer"
+#   | orElse (int "343")
+#   | orElse (int "asd")
+#   ---
+#   Some 343
+export def orElse (alternate: Option a): Option a => Option a = match _
+    Some a = Some a
+    None   = alternate
+
+# omap: apply function `f` to the optional contents
+# If you find yourself using the function with getOrElse, consider using a match instead.
+#
+#   omap (_+1) (Some 4) = Some 5
+#   omap (_+1) None     = None
+export def omap (f: a => b): Option a => Option b = match _
+    Some a = Some (f a)
+    None   = None
+
+# omapPartial: apply partial function 'f' to the optional contents
+# A partial function returns Option; only Some cases result in a value.
+#
+#   def isOdd x = x&1 == 1
```

Comment:
visual nit:
```
def isOdd x = x&1 == 1
```
has the `=` dividing the definition from the implementation but then the implementation has `==` on the same line. 
Just takes an extra second to mentally separate it out compared to
```
def isOdd x =
    x&1 == 1
```